Add `DiagramDetectConfig` to `DiagramsAPI.detect()`

ddonukis opened the pull request:
## Description
The diagram detect endpoint has a lot of configuration options that can be passed as a dict. But for that the user needs to know precisely which parameters are available and how they are spelled. To simplify configuring the detect job I've added a config data class. Of course the endpoint still supports raw dicts for backwards compatibility.

```python
from congite.client.data_classes.contextualization import DiagramDetectConfig, CustomizeFuzziness, ConnectionFlags

config = DiagramDetectConfig(
    read_embedded_text=True,
    customize_fuzziness=CustomizeFuzziness(
        min_chars=6,
    ),
    connection_flags=ConnectionFlags(natural_reading_order=True)
)

client.diagrams.detect(entities, file_ids, configuration=config)
```


## Checklist:
- [x] Tests added/updated.
- [x] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [x] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [x] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

olacognite reviewed (approved):
LGTM, mind that there is already a new parameter in town. Substitutions.

ddonukis commented:
> LGTM, mind that there is already a new parameter in town. Substitutions.

Added it. Also updated the param descriptions in line with the API spec PR.

erlendvollset commented on cognite/client/data_classes/contextualization.py:
```diff
@@ -1036,3 +1038,176 @@ def __init__(
     _RESOURCE = ResourceReference
+
+
+class NestableDiagramDetectConfig:
```
Let's make these classes extend `CogniteObject`.

ddonukis replied:
I answered in more detail in a different thread, but it doesn't seem like a good fit.

Although I just remembered the API returns this config as part of the detect job result 🤔 
Maybe we could load the results into this object then. I'll investigate further.

erlendvollset commented on tests/tests_unit/test_data_classes/test_contextualization.py:
```diff
@@ -181,3 +188,62 @@ def test_DetectJobBundle_div(self) -> None:
+            no_text_inbetween=True,
+            new_parameter=True,
+            excluded_parameter=False,
+            newCamelCaseParameter=True,
+        )
+        expected = [
+            "natural_reading_order",
+            "no_text_inbetween",
+            "new_parameter",
+            "newCamelCaseParameter",
+        ]
+        assert cf.dump() == expected
+
+    def test_nested_config(self) -> None:
+        config = DiagramDetectConfig(
+            annotation_extract=True,
+            case_sensitive=False,
+            connection_flags=ConnectionFlags(
+                natural_reading_order=False,
+            ),
+            direction_weights=DirectionWeights(
+                right=0.1,
+                down=0.1,
+            ),
+            min_fuzzy_score=0.95,
+            remove_leading_zeros=True,
+            new_parameter_str="something",
+            new_paramater_dict={"my_key": "my_value"},
+        )
+
+        expected = {
+            "annotationExtract": True,
+            "caseSensitive": False,
+            "directionWeights": {"right": 0.1, "down": 0.1},
+            "minFuzzyScore": 0.95,
+            "removeLeadingZeros": True,
+            "newParameterStr": "something",
+            "newParamaterDict": {"my_key": "my_value"},
+        }
+        assert config.dump() == expected
```
You'll get these tests generated when extending `CogniteObject`.

ddonukis replied:
Ohh! That's convenient.

ddonukis commented:
@erlendvollset made the config classes inherit from `CogniteObject` as you suggested 👌

ddonukis commented on cognite/client/_api/diagrams.py:
```diff
@@ -230,14 +237,31 @@ def detect(
+            config = configuration.dump() if isinstance(configuration, DiagramDetectConfig) else configuration
+            beta_parameters = dict(pattern_mode=pattern_mode, configuration=config)
+            self._detect_beta_params_warning.warn()
+            self._api_subversion = "beta"
```
@haakonvt does this look correct? I saw the datapoints endpoint appends `"-beta"` to subversion but setting subversion to `"beta"` seems to work for `digrams.detect`.

haakonvt replied:
The API versioning guide is very clear on what format it should be:
https://cognitedata.atlassian.net/wiki/spaces/PD/pages/3842900086/CDF+API+versioning

olacognite replied:
I think beta is equivalent to <latest>-beta, but better safe than sorry maybe

ddonukis replied:
Oh, thanks for sharing! I wasn't aware of that doc.

Will change to append '-beta'.

ddonukis replied:
Hmm, I'm not sure why the integration test is saying the api subversion is bad now 🤔

haakonvt replied:
@olacognite `Invalid version in cdf-version header`... 😉

haakonvt replied:
(Just double-check that it is in fact sending `"20230101-beta"`)

ddonukis commented on cognite/client/_api/diagrams.py:
```diff
@@ -230,14 +240,31 @@ def detect(
+            config = configuration.dump() if isinstance(configuration, DiagramDetectConfig) else configuration
+            beta_parameters = dict(pattern_mode=pattern_mode, configuration=config)
+            self._detect_beta_params_warning.warn()
+            self._api_subversion = "beta"
```
```suggestion
            if self._api_subversion and not self._api_subversion.endswith("-beta"):
                self._api_subversion += "-beta"
```

ddonukis commented:
Updated API subversion logic. Any other comments or can this be merged?

haakonvt commented on cognite/client/data_classes/contextualization.py:
```diff
@@ -1036,3 +1038,213 @@ def __init__(
+
+    @classmethod
+    def _load(cls, resource: dict[str, Any], cognite_client: CogniteClient | None = None) -> Self:
+        resouce_copy = resource.copy()
+
+        if con_flg := resouce_copy.pop("connectionFlags", None):
+            con_flg = ConnectionFlags.load(con_flg)
+        if cus_fuz := resouce_copy.pop("customizeFuzziness", None):
+            cus_fuz = CustomizeFuzziness.load(cus_fuz)
+        if dir_wgt := resouce_copy.pop("directionWeights", None):
+            dir_wgt = DirectionWeights.load(dir_wgt)
```
Perhaps we can change `pop` (mutates) to `get` (and skip initial copy)?
```suggestion
        if con_flg := resouce_copy.get("connectionFlags"):
            con_flg = ConnectionFlags.load(con_flg)
        if cus_fuz := resouce_copy.pop("customizeFuzziness"):
            cus_fuz = CustomizeFuzziness.load(cus_fuz)
        if dir_wgt := resouce_copy.pop("directionWeights"):
            dir_wgt = DirectionWeights.load(dir_wgt)
```

haakonvt commented on cognite/client/data_classes/contextualization.py:
```diff
@@ -1036,3 +1038,213 @@ def __init__(
+        if dir_wgt := resouce_copy.pop("directionWeights", None):
+            dir_wgt = DirectionWeights.load(dir_wgt)
+
+        snake_cased = convert_all_keys_to_snake_case(resouce_copy)
```
Perhaps just pass these on without any conversion?

haakonvt commented on pyproject.toml:
```diff
@@ -1,7 +1,7 @@
 name = "cognite-sdk"
 
-version = "7.39.0"
+version = "7.39.1"
```
I think this deserves a minor bump 👌

ddonukis commented on cognite/client/data_classes/contextualization.py:
```diff
@@ -1036,3 +1038,213 @@ def __init__(
+    larger distance is allowed.
+
+    Args:
+        left (float | None): explanation
+        right (float | None): explanation
+        up (float | None): explanation
+        down (float | None): explanation
```
not self explanatory enough? 😜

yeah, I forgot about it

ddonukis commented on cognite/client/data_classes/contextualization.py:
```diff
@@ -1036,3 +1038,213 @@ def __init__(
+        )
+
+
+class CustomizeFuzziness(CogniteObject):
```
It's a bit of an unfortunate name, straight from the context api.

ddonukis replied:
I thought about changing the name and I believe it's better to keep it consistent with the API for the sake of user experience. Now the name matches the API field exactly so it's easy to look it up in the docs and code.

The name really should changed on the API side which talked to Ola about. He doesn't like the name either.

haakonvt replied:
Ah, I see. This still has to be passed as `customize_fuzziness=CustomizeFuzziness(...)` so I think you are free to choose a better name for it 😄 

Maybe @erlendvollset thinks otherwise

ddonukis commented:
@haakonvt  thanks for the review, I'll fix next week 👍

ddonukis commented on cognite/client/data_classes/contextualization.py:
```diff
@@ -1036,3 +1038,213 @@ def __init__(
+    def dump(self, camel_case: bool = True) -> dict[str, Any]:
+        dumped = super().dump(camel_case=camel_case)
+
+        for param_name, v in self._extra_params.items():
+            dumped[to_camel_case(param_name) if camel_case else param_name] = v
```
I don't understand what you mean. I'm just respecting the `camel_case` param. It's true by default anyways. If for some reason the user wants to dump the object in snake case I think it's more weird that some of the config options are converted and some are not.

I'm thinking that the user can stick to the pythonic snake_case when using the config class but could also use camelCase, so we don't really know which case the use will provide when initializing the class.

When loading we expect the values to come from the api, in camelCase, since it's not arbitrary metadata but part of the api schema, albeit not documented.

haakonvt replied:
All I'm trying to say is that a roundtrip using `to_camel_case` and `to_snake_case` can only be considered a no-op for well-behaved strings, that's why I think you should not pass them through either of these functions, not in load and not in dump.

If a user is passing these super-custom settings, they need to account for CorrectCasing themselves. Just my 2 cents (I'll leave the decision up to you!)

haakonvt replied:
(A bit like how we in data modeling aren't case converting properties)

ddonukis commented on cognite/client/data_classes/contextualization.py:
```diff
@@ -1036,3 +1038,212 @@ def __init__(
+
+    @classmethod
+    def _load(cls, resource: dict[str, Any], cognite_client: CogniteClient | None = None) -> Self:
+        resource_copy = resource.copy()
```
@haakonvt I do the copy and pop because I expect undocumented parameters to be in this resource dict.

If I use get and then pass the original dict into kwargs, the parsed and loaded params like `ConnectionFlags` will be passed to `cls()` twice, causing a TypeError.

I need some way to separate known loadable params from unknown ones, so I decided to use `pop()` to remove the items we need to load separately.

ddonukis commented:
@haakonvt I fixed the subversion handling and the failing test (it was modifying subversion separately, so api received subversion `"beta-beta"`). Also addressed you other comments.

I'm pretty happy with how the PR looks now. What do you think?

haakonvt commented on cognite/client/_api/diagrams.py:
```diff
@@ -230,14 +243,34 @@ def detect(
+            config = configuration.dump() if isinstance(configuration, DiagramDetectConfig) else configuration
+            beta_parameters = dict(pattern_mode=pattern_mode, configuration=config)
+
+            self._detect_beta_params_warning.warn()
+            if self._api_subversion and not self._api_subversion.endswith("beta"):
+                api_subversion = f"{self._api_subversion}-beta"
```
Is this calling `beta` even when no beta features are used?

haakonvt replied:
...or is passing `configuration` always beta perhaps

haakonvt reviewed (approved):
LGTM!
